tcms/testruns/data.py

```diff
@@ -88,31 +97,38 @@ def get_case_runs_comments(run_pk):
         @return: the mapping between case run id and comments
         @rtype: dict
         """
-        content_type = ContentType.objects.get_for_model(TestCaseRun)
-
-        rows = []
         # note: cast to string b/c object_pk is a Textield and PostgreSQL
         # doesn't like TEXT in <list of integers>
         # in Django 1.10 we have the Cast() function for similar cases, see
         # https://docs.djangoproject.com/en/1.10/ref/models/database-functions/#cast
-        obj_pks = [str(o.pk) for o in TestCaseRun.objects.filter(run=run_pk).only('pk')]
-        for row in Comment.objects.filter(
-                site=settings.SITE_ID,
-                content_type=content_type.pk,
-                is_public=True,
-                is_removed=False,
-                object_pk__in=obj_pks
-            ).annotate(
-                case_run_id=F('object_pk')
-            ).values(
-                'case_run_id',
-                'submit_date',
-                'comment',
-                'user_name').order_by('case_run_id'):
+        object_pks = []
+        for test_case_run in TestCaseRun.objects.filter(run=run_pk).only('pk'):
+            object_pks.append(str(test_case_run.pk))
+
+        comments = Comment.objects.filter(
+            site=settings.SITE_ID,
+            content_type=ContentType.objects.get_for_model(TestCaseRun).pk,
+            is_public=True,
+            is_removed=False,
+            object_pk__in=object_pks
+        ).annotate(
```

Comment:
Parts of this method are duplicate with the `get_comments()` function. I think it should be possible to use get_comments() and then perform additional filtering and annotation on the resulting query set. Give it a try.

Reply:
Actually, as far as I can see `get_comments()` works in a different way, the main difference being that it gets the comment for a single object, while here we get the comments for a list of them. Even if we refactor this method to use `get_comments()` and make `get_comments()` more customisable, by passing more arguments to it (right now we pass just an `obj` and it would not work in our case), this would result in increasing number of SQL queries.

Reply:
Fine for now, but we should seek to re-use existing methods as much as possible. I still don't quite like the fact that this piece of code duplicates existing functionality, although in a more extended way.

Reply:
Me too. But to reuse the method now, it means that we have to modify it, with the risk of breaking something. Still, it has to be done at some point.